hibike/hibike_process.py

```diff
@@ -85,25 +90,124 @@ def recv_subscription_requests(conn, uid_queue, stop_event):
+    pack.read_thread = threading.Thread(target=device_read_thread,
+                                        args=(uid, serial_port, pack.write_queue, error_queue,
+                                              state_queue, batched_data))
+    pack.write_thread.start()
+    pack.read_thread.start()
+    return pack
+
+
+def hotplug(devices, state_queue, batched_data, error_queue):
+    """
+    Remove disconnected devices and scan for new ones.
+    """
+    clean_up_queue = queue.Queue()
+    clean_up_thread = threading.Thread(target=clean_up_devices, args=(clean_up_queue, ))
+    clean_up_thread.start()
+    while True:
+        time.sleep(HOTPLUG_POLL_INTERVAL)
+        scan_for_new_devices(devices, state_queue, batched_data, error_queue)
+        remove_disconnected_devices(error_queue, devices, clean_up_queue, state_queue)
+
+
+def scan_for_new_devices(existing_devices, state_queue, batched_data, error_queue):
+    """
+    Find devices that are on serial ports not in EXISTING_DEVICES, and add
+    any that have been found to it.
+    """
+    ports, names = get_working_serial_ports(map(lambda d: d.serial_port.name,
+                                                existing_devices.values()))
+    sensors = identify_smart_sensors(ports)
+    for (ser, uid) in sensors.items():
+        idx = names.index(ser)
+        port = ports[idx]
+        pack = spin_up_device(port, uid, state_queue, batched_data, error_queue)
+        existing_devices[uid] = pack
+        # Tell the device to start sending data
+        pack.write_queue.put(("ping", []))
+        pack.write_queue.put(("subscribe", [1, 0, []]))
+
+
+def clean_up_devices(device_queue):
+    """
+    Clean up associated resources of devices in the queue.
+
+    Closing a serial port can take a very long time (30 seconds or more).
+    It's best to spin this function off into its own thread,
+    so that you're not blocked on reclaiming resources.
+    """
+    while True:
+        device = device_queue.get()
+        device.serial_port.close()
+        device.read_thread.join()
+        device.write_thread.join()
+
+
+def remove_disconnected_devices(error_queue, devices, clean_up_queue, state_queue):
+    """
+    Clean up any disconnected devices in ERROR_QUEUE.
+    """
+    next_time_errors = []
+    while True:
+        try:
+            error = error_queue.get(block=False)
+            pack = devices[error.uid]
+            if not error.accessed:
+                # Wait until the next cycle to make sure it's disconnected
+                error.accessed = True
+                next_time_errors.append(error)
+                continue
+            elif error.serial_port.name != pack.serial_port.name:
+                # We found it again
+                continue
+            uid = error.uid
+            pack = devices[uid]
+            del devices[uid]
+            clean_up_queue.put(pack)
+            state_queue.put(("device_disconnected", [uid]))
+        except queue.Empty:
```

Comment:
Is there a chance that the queue will never be empty, because something else keeps adding to it?
if so, nothing will ever get put on the error_queue, correct?